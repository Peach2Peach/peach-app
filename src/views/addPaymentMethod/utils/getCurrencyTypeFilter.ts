export const getCurrencyTypeFilter = (type: 'europe' | 'other') => (currency: Currency) => {
  if (type === 'europe') return currency !== 'USDT'
  return currency === 'USDT'
}
