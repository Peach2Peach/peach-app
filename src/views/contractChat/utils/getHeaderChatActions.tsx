import React from 'react'
import { Icon } from '../../../components'
import { ConfirmCancelTrade } from '../../../overlays/ConfirmCancelTrade'
import { ConfirmRaiseDispute } from '../../../overlays/ConfirmRaiseDispute'
import tw from '../../../styles/tailwind'
import { HeaderConfig } from '../../../components/header/store'

// eslint-disable-next-line max-len
export const getHeaderChatActions = (
  contract: Contract,
  view: 'buyer' | 'seller' | '',
  updateOverlay: React.Dispatch<OverlayState>,
): HeaderConfig['icons'] => {
  const canCancel
    = !contract.disputeActive && !contract.paymentMade && !contract.canceled && !contract.cancelationRequested
  const canDispute
    = contract.symmetricKey
    && ((!contract.disputeActive && !/cash/u.test(contract.paymentMethod))
      || (view === 'seller' && contract.cancelationRequested))

  const openCancelTrade = () =>
    canCancel
      ? updateOverlay({
        content: <ConfirmCancelTrade contract={contract} />,
        visible: true,
      })
      : null
  // const extendTime = () => alert('todo extend time')
  const raiseDispute = () =>
    canDispute
      ? updateOverlay({
        content: <ConfirmRaiseDispute contract={contract} />,
        visible: true,
      })
      : null

  const icons: HeaderConfig['icons'] = [
    {
      iconComponent: <Icon style={!canCancel && tw`opacity-50`} id="xCircle" color={tw`text-error-main`.color} />,
      onPress: openCancelTrade,
    },
    {
      iconComponent: (
        <Icon style={!canDispute && tw`opacity-50`} id="alertOctagon" color={tw`text-warning-main`.color} />
      ),
      onPress: raiseDispute,
    },
  ]
  return icons
}
