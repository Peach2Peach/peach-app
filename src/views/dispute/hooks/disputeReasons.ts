export const disputeReasons: Record<string, DisputeReason[]> = {
  buyer: ['noPayment.buyer', 'unresponsive.buyer', 'abusive', 'other'],
  seller: ['noPayment.seller', 'unresponsive.seller', 'abusive', 'other'],
}
