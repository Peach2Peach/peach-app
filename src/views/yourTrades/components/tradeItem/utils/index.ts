export { getActionIcon } from './getActionIcon'
export { getActionLabel } from './getActionLabel'
