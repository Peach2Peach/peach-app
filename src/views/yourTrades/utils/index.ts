export { getDisputeResultTheme } from './getDisputeResultTheme'
export { getNavigationDestinationForContract, getNavigationDestinationForOffer } from './getNavigationDestination'
export { getOfferLevel } from './getOfferLevel'
export { getThemeForPastTrade } from './getThemeForPastTrade'
export { hasDoubleMatched } from './hasDoubleMatched'
export { isContractSummary } from './isContractSummary'
export { isOpenAction } from './isOpenAction'
export { isOpenOffer } from './isOpenOffer'
export { isPastOffer } from './isPastOffer'
export { isPrioritary } from './isPrioritary'
export { isWaiting } from './isWaiting'
export { navigateToContract } from './navigateToContract'
export { navigateToOffer } from './navigateToOffer'
export { shouldGoToOfferSummary } from './shouldGoToOfferSummary'
export { statusIcons } from './statusIcons'
export { wonDisputeForTrade } from './wonDisputeForTrade'
