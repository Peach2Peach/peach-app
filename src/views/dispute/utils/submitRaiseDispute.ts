import { useConfigStore } from '../../../store/configStore'
import { error } from '../../../utils/log'
import { peachAPI } from '../../../utils/peachAPI'
import { signAndEncrypt } from '../../../utils/pgp'

export const submitRaiseDispute = async (
  contract: Contract | undefined,
  reason: DisputeReason,
  email?: string,
  message?: string,
  // eslint-disable-next-line max-params
): Promise<[boolean, APIError | null]> => {
  if (!contract || !contract.symmetricKey) return [false, null]
  const { encrypted: symmetricKeyEncrypted } = await signAndEncrypt(
    contract.symmetricKey,
    useConfigStore.getState().peachPGPPublicKey,
  )
  const { result, error: err } = await peachAPI.private.contract.raiseDispute({
    contractId: contract.id,
    email,
    reason,
    message,
    symmetricKeyEncrypted,
  })
  if (result) {
    return [true, null]
  }
  if (err) {
    error('Error', err)
    return [false, err]
  }
  return [false, null]
}
