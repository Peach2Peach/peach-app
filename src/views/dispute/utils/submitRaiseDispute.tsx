import { configStore } from '../../../store/configStore'
import { account } from '../../../utils/account'
import { getChat, saveChat } from '../../../utils/chat'
import { initDisputeSystemMessages } from '../../../utils/chat/initDisputeSystemMessages'
import { error } from '../../../utils/log'
import { raiseDispute } from '../../../utils/peachAPI'
import { signAndEncrypt } from '../../../utils/pgp'

export const submitRaiseDispute = async (
  contract: Contract | undefined,
  reason: DisputeReason,
  email?: string,
  message?: string,
  // eslint-disable-next-line max-params
): Promise<[boolean, APIError | null]> => {
  if (!contract || !contract.symmetricKey) return [false, null]
  const { encrypted: symmetricKeyEncrypted } = await signAndEncrypt(
    contract.symmetricKey,
    configStore.getState().peachPGPPublicKey,
  )
  const [result, err] = await raiseDispute({
    contractId: contract.id,
    email,
    reason,
    message,
    symmetricKeyEncrypted,
  })
  if (result) {
    const updatedContract = {
      ...contract,
      disputeDate: new Date(Date.now()),
      disputeInitiator: account.publicKey,
      disputeReason: reason,
    }
    const chat = getChat(contract.id)
    const autogeneratedMessages = initDisputeSystemMessages(chat.id, updatedContract)
    saveChat(contract.id, {
      messages: autogeneratedMessages,
    })
    return [true, null]
  }
  if (err) {
    error('Error', err)
    return [false, err]
  }
  return [false, null]
}
