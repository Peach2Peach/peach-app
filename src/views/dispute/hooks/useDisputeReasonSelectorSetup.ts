import { useHeaderSetup, useNavigation, useRoute } from '../../../hooks'
import { useShowErrorBanner } from '../../../hooks/useShowErrorBanner'
import { useDisputeRaisedSuccess } from '../../../popups/dispute/hooks/useDisputeRaisedSuccess'
import { account } from '../../../utils/account'
import { contractIdToHex, getContract, getContractViewer } from '../../../utils/contract'
import i18n from '../../../utils/i18n'
import { submitRaiseDispute } from '../utils/submitRaiseDispute'
import { disputeReasons } from './disputeReasons'

export const useDisputeReasonSelectorSetup = () => {
  const route = useRoute<'disputeReasonSelector'>()
  const contract = getContract(route.params.contractId)

  const view = contract ? getContractViewer(contract, account) : ''
  const availableReasons = view === 'seller' ? disputeReasons.seller : disputeReasons.buyer

  const navigation = useNavigation()
  const showErrorBanner = useShowErrorBanner()

  const showDisputeRaisedPopup = useDisputeRaisedSuccess()

  useHeaderSetup(i18n('dispute.disputeForTrade', !!contract ? contractIdToHex(contract.id) : ''))

  const setAndSubmit = async (reason: DisputeReason) => {
    const [success, error] = await submitRaiseDispute(contract, reason)
    if (!success || error) {
      showErrorBanner(error?.error)
      return
    }
    if (view) showDisputeRaisedPopup(view)
    navigation.goBack()
  }

  const setReason = async (reason: DisputeReason) => {
    if (!contract) return
    if (reason === 'noPayment.buyer' || reason === 'noPayment.seller') {
      navigation.navigate('disputeForm', { contractId: contract.id, reason })
      return
    }

    await setAndSubmit(reason)
  }

  return { availableReasons, setReason }
}
