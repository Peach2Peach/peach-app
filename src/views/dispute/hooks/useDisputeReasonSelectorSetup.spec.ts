import { renderHook } from '@testing-library/react-native'
import { account1 } from '../../../../tests/unit/data/accountData'
import { contract } from '../../../../tests/unit/data/contractData'
import { setAccount } from '../../../utils/account'
import i18n from '../../../utils/i18n'
import { disputeReasons } from './disputeReasons'
import { useDisputeReasonSelectorSetup } from './useDisputeReasonSelectorSetup'
import { apiSuccess, unauthorizedError } from '../../../../tests/unit/data/peachAPIData'

const useRouteMock = jest.fn().mockReturnValue({
  params: {
    contractId: contract.id,
  },
})
jest.mock('../../../hooks/useRoute', () => ({
  useRoute: () => useRouteMock(),
}))

const navigateMock = jest.fn()
const goBackMock = jest.fn()
jest.mock('../../../hooks/useNavigation', () => ({
  useNavigation: () => ({
    navigate: navigateMock,
    goBack: goBackMock,
  }),
}))

const useHeaderSetupMock = jest.fn()
jest.mock('../../../hooks/useHeaderSetup', () => ({
  useHeaderSetup: (...args: any[]) => useHeaderSetupMock(...args),
}))

const getContractMock = jest.fn().mockReturnValue(contract)
jest.mock('../../../utils/contract/getContract', () => ({
  getContract: (...args: any[]) => getContractMock(...args),
}))

const submitRaiseDisputeMock = jest.fn().mockResolvedValue([apiSuccess, null])
jest.mock('../utils/submitRaiseDispute', () => ({
  submitRaiseDispute: (...args: any[]) => submitRaiseDisputeMock(...args),
}))

const showErrorBannerMock = jest.fn()
jest.mock('../../../hooks/useShowErrorBanner', () => ({
  useShowErrorBanner: () => showErrorBannerMock,
}))
const showDisputeRaisedPopupMock = jest.fn()
jest.mock('../../../popups/dispute/hooks/useDisputeRaisedSuccess', () => ({
  useDisputeRaisedSuccess: () => showDisputeRaisedPopupMock,
}))

describe('useDisputeReasonSelectorSetup', () => {
  it('returns default values correctly for seller', () => {
    setAccount({ ...account1, publicKey: contract.seller.id })
    const { result } = renderHook(useDisputeReasonSelectorSetup)

    expect(result.current.availableReasons).toEqual(disputeReasons.seller)
    expect(result.current.setReason).toBeInstanceOf(Function)
  })
  it('returns default values correctly for buyer', () => {
    setAccount({ ...account1, publicKey: contract.buyer.id })
    const { result } = renderHook(useDisputeReasonSelectorSetup)

    expect(result.current.availableReasons).toEqual(disputeReasons.buyer)
  })
  it('respects route params', () => {
    useRouteMock.mockReturnValueOnce({
      params: {
        contractId: contract.id,
      },
    })
    renderHook(useDisputeReasonSelectorSetup)
    expect(getContractMock).toHaveBeenCalledWith(contract.id)
  })
  it('sets up the header correctly', () => {
    renderHook(useDisputeReasonSelectorSetup)
    expect(useHeaderSetupMock).toHaveBeenCalledWith({
      title: i18n('dispute.disputeForTrade', 'PC‑E‑F'),
    })
  })
  it('does not set reason if no contract could be fetched', async () => {
    getContractMock.mockReturnValueOnce(undefined)
    const { result } = renderHook(useDisputeReasonSelectorSetup)
    await result.current.setReason('other')
    expect(submitRaiseDisputeMock).not.toHaveBeenCalled()
    expect(navigateMock).not.toHaveBeenCalled()
  })
  it('sets reason and navigate to dispute form if reason is noPayment.buyer', async () => {
    const reason = 'noPayment.buyer'
    const { result } = renderHook(useDisputeReasonSelectorSetup)
    await result.current.setReason(reason)
    expect(navigateMock).toHaveBeenCalledWith('disputeForm', { contractId: contract.id, reason })
    expect(showDisputeRaisedPopupMock).not.toHaveBeenCalled()
  })
  it('sets reason and navigate to dispute form if reason is noPayment.seller', async () => {
    const reason = 'noPayment.seller'
    const { result } = renderHook(useDisputeReasonSelectorSetup)
    await result.current.setReason(reason)
    expect(navigateMock).toHaveBeenCalledWith('disputeForm', { contractId: contract.id, reason })
    expect(showDisputeRaisedPopupMock).not.toHaveBeenCalled()
  })
  it('sets reason and submits dispute request if reason is any other', async () => {
    const reason = 'abusive'
    const { result } = renderHook(useDisputeReasonSelectorSetup)
    await result.current.setReason(reason)
    expect(submitRaiseDisputeMock).toHaveBeenCalledWith(contract, reason)
    expect(showDisputeRaisedPopupMock).toHaveBeenCalled()
  })
  it('shows error banner if dispute request fails', async () => {
    submitRaiseDisputeMock.mockResolvedValueOnce([null, unauthorizedError])
    const reason = 'abusive'
    const { result } = renderHook(useDisputeReasonSelectorSetup)
    await result.current.setReason(reason)
    expect(showErrorBannerMock).toHaveBeenCalledWith(unauthorizedError.error)
    expect(showDisputeRaisedPopupMock).not.toHaveBeenCalled()
  })
})
