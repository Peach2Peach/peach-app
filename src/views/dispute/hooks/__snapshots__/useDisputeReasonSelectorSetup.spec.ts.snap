// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDisputeReasonSelectorSetup sets up the header correctly 1`] = `<React.Fragment />`;
